sessions/forecast-ensembles.qmd

```diff
@@ -48,127 +52,230 @@ The source file of this session is located at `sessions/forecast-ensembles.qmd`.
+
+```{r}
+sampled_forecasts <- generate(
+  cv_models, 
+  h=8, 
+  times=100) |> 
+  group_by(.split, region, .model, .rep) |> 
+  mutate(h = row_number())
+sampled_forecasts
+```
+
+These forecasts above are still in `fable`-style formatting.
+The code below renames and reformats columns and makes this valid hubverse-style model output.
+
+```{r convert-for-hubverse}
+sampled_forecasts_hub <- sampled_forecasts |> 
+  rename(value = .sim,
+         model_id = .model,
+         target_epiweek = epiweek) |> 
+  mutate(output_type = "sample",
+         output_type_id = stringr::str_c(region, stringr::str_pad(.rep, width = 3, pad = "0")),
+         forecast_date = target_epiweek - h*7L) |> 
+  ungroup() |> 
+  as_tibble() |> 
+  select(model_id,
+         forecast_date, 
+         target_epiweek,
+         h,
+         output_type,
+         output_type_id,
+         value)
+sampled_forecasts_hub
+```
+
+One of the foundational data structures for hubverse-style output is the three columns:
+
+-   `output_type`: here we have `sample` trajectories
+-   `output_type_id`: for `sample` output type, this column contains unique IDs that identify, for each `forecast_date`, which rows are from the same trajectory.
+-   `value`: this is the actual predicted value.
+
+We can quickly visualize these sample trajectory outputs to see the structure of the forecasts for one week and one model:
+
+```{r}
+sampled_forecasts_hub |> 
+  filter(forecast_date == as.Date("2018-01-14"),
+         model_id == "fourier_ar2") |> 
+  ggplot() +
+  geom_line(aes(x = target_epiweek, y = value, group = output_type_id), 
+            color = "red",
+            alpha = 0.2) +
+  geom_line(data = flu_data |> filter(epiweek >= as.Date("2017-09-01")), 
+            aes(x=epiweek, y=wili))
+```
+
 # Forecast ensembles
 
-We will now move to creating forecasts as a combination of multiple forecasts.
-This procedure is also sometimes called _stacking_, and the resulting forecasts are said to come from _ensembles_ of forecast models.
+Now that we have standardized our data formatting, we will now move to creating forecasts as a combination of multiple forecasts.
+This procedure is also sometimes called *stacking*, and the resulting forecasts are said to come from *ensembles* of forecast models.
 
 ## Quantile-based ensembles
 
 We will first consider forecasts based on the individual quantiles of each model.
 This corresponds to a situation where each forecast aims to correctly determine a single target predictive distribution.
 By taking an average of all models, we aim to get a better estimate of this distribution than from the individual models.
-If we have reason to believe that some models are better than others at estimating this distribution, we can create a weighted version of this average.
+If we have reason to believe that some models are **consistently** better than others at estimating this distribution, we could try to create a weighted version of this average.
```

Comment:
I don't think this is quite correct. We can always create a weighted version its just that it might not be any good. Suggest an edit.